src/main/java/org/elasticsearch/action/admin/indices/upgrade/post/TransportUpgradeAction.java

```diff
@@ -0,0 +1,214 @@
+                                  TransportUpgradeSettingsAction upgradeSettingsAction) {
+        super(settings, UpgradeAction.NAME, threadPool, clusterService, transportService, actionFilters,
+                UpgradeRequest.class, ShardUpgradeRequest.class, ThreadPool.Names.OPTIMIZE);
+        this.indicesService = indicesService;
+        this.upgradeSettingsAction = upgradeSettingsAction;
+    }
+
+    @Override
+    protected UpgradeResponse newResponse(UpgradeRequest request, AtomicReferenceArray shardsResponses, ClusterState clusterState) {
+        int successfulShards = 0;
+        int failedShards = 0;
+        List<ShardOperationFailedException> shardFailures = null;
+        Map<String, Integer> successfulPrimaryShards = newHashMap();
+        Map<String, Version> versions = newHashMap();
+        for (int i = 0; i < shardsResponses.length(); i++) {
+            Object shardResponse = shardsResponses.get(i);
+            if (shardResponse == null) {
+                // a non active shard, ignore...
+            } else if (shardResponse instanceof BroadcastShardOperationFailedException) {
+                failedShards++;
+                if (shardFailures == null) {
+                    shardFailures = newArrayList();
+                }
+                shardFailures.add(new DefaultShardOperationFailedException((BroadcastShardOperationFailedException) shardResponse));
+            } else {
+                successfulShards++;
+                ShardUpgradeResponse shardUpgradeResponse = (ShardUpgradeResponse) shardResponse;
+                String index = shardUpgradeResponse.getIndex();
+                if (shardUpgradeResponse.primary()) {
+                    Integer count = successfulPrimaryShards.get(index);
+                    successfulPrimaryShards.put(index, count == null ? 1 : count + 1);
+                }
+                Version version = versions.get(index);
+                if (version == null || shardUpgradeResponse.version().onOrAfter(version) == false) {
+                    versions.put(index, shardUpgradeResponse.version());
+                }
+            }
+        }
+        Map<String, String> updatedVersions = newHashMap();
+        MetaData metaData = clusterState.metaData();
+        for (Map.Entry<String, Version> versionEntry : versions.entrySet()) {
+            String index = versionEntry.getKey();
+            Integer primaryCount = successfulPrimaryShards.get(index);
+            int expectedPrimaryCount = metaData.index(index).getNumberOfShards();
+            if (primaryCount == metaData.index(index).getNumberOfShards()) {
+                updatedVersions.put(index, versionEntry.getValue().toString());
+            } else {
+                logger.warn("Not updating settings for the index [{}] because upgraded of some primary shards failed - expected[{}], received[{}]", index,
```

Comment:
This is tricky. It could have been not upgraded because somebody restarted a node during upgrade procedure. I think simply not upgrading the index if we cannot upgrade all shards would be a safer solution here.